teps/0017-shell-escaped-params.md

```diff
@@ -0,0 +1,151 @@
+---
+title: Provide Shell-Escaped Parameters
+authors:
+  - "@coryrc"
+creation-date: 2020-09-15
+last-updated: 2020-09-15
+status: proposed
+---
+
+# TEP-0017 Provide Shell-Escaped Parameters
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+It is difficult to safely use parameters in a `script:` scalar. This TEP proposes
+offering a shell-escaped version of every parameter.
+
+## Motivation
+
+- More concise than the current workarounds
+- Help users create more robust Tasks by default
+- More clearly expresses intent
+
+### Goals
+
+Provide a way to place parameters directly into the default `script:` without being
+interpreted by the shell.
+
+### Non-Goals
+
+1. Prevent purposeful interpretation of parameters when intended
+2. Work with shells which are not the default for the `script:`
+3. Be perfectly secure against arbitrary data
+
+## Requirements
+
+## Proposal
+
+In a Task or Pipeline, for each parameter `foo`, provide `$(params.foo.shell-escaped)`.
+
+This variable shall be escaped according to `printf '%q'` rules as used in bash.
+For example, the literal string `Hello '\!"` becomes `Hello\ \'\\\!\"`. A script
+which ran `echo $(params.foo.shell-escaped)` would print the literal string. The
+escaped value would do the wrong thing when placed inside double or single quotes.
+
+Support Bourne, Bourne-Again, Busybox, and Debian Almquist shells against
+injection, but don't purposefully prevent its usage in other scenarios. This
+does not require runtime analysis.
+
+### User Stories
+
+#### Story 1
+
+My Task allows users to specify a destination filename. The user could pass
+valid file names containing any of ` <>|\!*?;&"'` (or more) which would be
+impossible to robustly address all possibilities when used directly in the
+script. For example, single-quoting parameters could not protect against
+`'; rm -fR *;'`.
+
+#### Story 2
+
+My Task has a `script:` to perform some function and I wish the user to be able
+to provide their own script to be called by my script. The most concise way to
+write my Task is to include the line:
+
+```
+printf '%s' $(params.foo.shell-escaped) > included-script.sh
+```
+
+### Risks and Mitigations
+
+We may not have the resources to exhaustively test and security-harden this feature.
+
+## Design Details
+
+Arrays will be treated as flattened strings.
```

Comment:
if we go ahead, what do you think about leaving this out? one reason we haven't provided a feature to flatten arrays as strings so far is it raises a lot of questions, e.g. for starters are the values in the resulting string separated by spaces or commas or other?

Reply:
Something will happen if it's an array -- I'm not sure what that something should be, but I think it should be defined.

Reply:
i suggest for now we fail validation if you try to use `.shell-escaped` on an array

Reply:
done